tiledb/api/c_api/group/group_api_internal.h

```diff
@@ -0,0 +1,84 @@
+/**
+ * @file tiledb/api/c_api/group/group_api_internal.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2022 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file declares the internal group section of the C API for TileDB.
+ */
+
+#ifndef TILEDB_CAPI_GROUP_API_INTERNAL_H
+#define TILEDB_CAPI_GROUP_API_INTERNAL_H
+
+#include "../../c_api_support/handle/handle.h"
+#include "../error/error_api_internal.h"
+#include "tiledb/sm/group/group.h"
+#include "tiledb/sm/group/group_v1.h"
+
+struct tiledb_group_handle_t
+    : public tiledb::api::CAPIHandle<tiledb_group_handle_t> {
+  /**
+   * Type name
+   */
+  static constexpr std::string_view object_type_name{"group"};
+
+ private:
+  tiledb::sm::GroupV1 group_;
+
+ public:
+  tiledb_group_handle_t() = delete;
+
+  explicit tiledb_group_handle_t(
+      const tiledb::sm::URI& uri, tiledb::sm::StorageManager* sm)
+      : group_(uri, sm) {
+  }
+
+  [[nodiscard]] inline tiledb::sm::Group& group() {
+    return group_;
+  }
+
+  [[nodiscard]] inline tiledb::sm::Group& group() const {
+    // Pretty sure this is gross and maybe broken, but it unsticks me for now.
+    // return group_;
```

Comment:
Compiler error for this:

```
error: binding reference of type 'tiledb::sm::Group' to value of type 'const tiledb::sm::GroupV1' drops 'const' qualifier
```

If you meant `const_cast` instead of `static_cast` it gives this error:

```
error: const_cast from 'const tiledb::sm::GroupV1' to 'tiledb::sm::Group &' is not allowed
```

I can get it to compile by casting away const'ness before static casting to the base class though:

```c++
    return static_cast<tiledb::sm::Group&>(
        const_cast<tiledb::sm::GroupV1&>(group_));
```

I've got no idea which version to prefer here though.

Reply:
I didn't notice the `const` qualifier. The double cast is the right way to express this situation, since there's both a type cast and a const cast.

Reply:
That's fair. Now fixed.